waspc/src/Wasp/AppSpec/Valid.hs

```diff
@@ -49,6 +58,49 @@ validateExactlyOneAppExists spec =
         GenericValidationError $
           "You have more than one 'app' declaration in your Wasp app. You have " ++ show (length apps) ++ "."
 
+validateWasp :: AppSpec -> [ValidationError]
+validateWasp spec = validateWaspVersion specWaspVersionStr
+  where
+    specWaspVersionStr = Wasp.version $ App.wasp (snd $ getApp spec)
+
+validateWaspVersion :: String -> [ValidationError]
+validateWaspVersion specWaspVersionStr = eitherUnitToErrorList $ do
+  specWaspVersionRange <- parseWaspVersionRange specWaspVersionStr
+  unless
+    ( SV.isVersionInRange currentWaspVersion specWaspVersionRange
+    )
+    $ Left $ incompatibleVersionError currentWaspVersion specWaspVersionRange
```

Comment:
You can use list comprehensions to avoid wrapping in and out of Either and return a list of errors directly:
```haskell
validateWaspVersion :: String -> [ValidationError]
validateWaspVersion specWaspVersionStr = case parseWaspVersionRange specWaspVersionStr of
  Left err -> [err]
  Right versionRange ->
    [ incompatibleVersionError currentWaspVersion versionRange
      | not $ SV.isVersionInRange currentWaspVersion versionRange
    ]
  where
    ...
```
Doing so will allow you to remove the `eitherUnitToErrorList` function.

Reply:
Thank you for reviewing my PR :smile: 

The use of `eitherUnitToErrorList` comes from [the review by @Martinsos ](https://github.com/wasp-lang/wasp/pull/786#discussion_r1006936789) . Personally, I like more wrapping it in an `Either` as it seems more appropriate for a validation function, but I'm not sure which one is more readable. What do you think?

Reply:
> Personally, I like more wrapping it in an Either as it seems more appropriate for a validation function

I'd normally agree, but in this case we only wrap the `Either` value to unwrap it (i.e., we get no benefits from using the Either type, only overhead). In other words, this function is the validation function and is set to return a list of errors. I don't think we're gaining much by implementing its body in terms of Either if we aren't using it for anything besides immediately converting it into a list. Plus, it adds extra complexity for the reader (i.e., an additional local function, a `do` block, a call to `unless`).
Moreover, the part of the code that uses `Either` isn't even a local function (i.e., inside a where clause), it's completely inline.

All in all, unless I'm missing something, I'd just directly construct a list. @Martinsos can correct me if I am.

Reply:
You're right. It does add to the complexity of the function. I'm going to refactor it :D

Reply:
@sodic the approach with Either was suggested by me, since it seemed to me that we could benefit from short-circuiting that it offers as a Monad. Validation happens in multiple steps, and can fail in any of those, so it is a good fit for short-circuiting. In this case, it is only two steps, but even with that we get to avoid nesting the second check (isVersionInRange) into the resolution of first step (parsing), so I would say that is a benefit of this approach.

I didn't understand the part with `where` and local function.

I would argue that extra local function is very descriptive and doesn't really affect the reader.

I agree that using monads and therefore do and unless are a bit more complex for the beginner, but we are writing our codebase with assumption that readers do have at least basic understanding of monads, so I wouldn't worry about that.

At the end I think both approaches are ok, the Either one and one you suggested -> I like Either a bit more due to no nesting and ability to expand by adding more steps easily if needed, but I do like that with your approach there is no need for that local function (although that doesn't break the deal for me).

Reply:
Well, @NeoLight1010, I guess you can choose whichever version you like better then :smile:

Reply:
Yup, I guess it's a matter of personal preference at this point. I'll leave the `Either` one for now. Thank you for your feedback :D